### PR TITLE
Lazy mask_fillvalues preprocessor function

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Make the [`mask_fillvalues`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html#missing-values-masks) preprocessor function lazy.

Note that this changes how provenance is recorded: now all cubes/datasets that are used as input to the function are considered ancestors of the output cubes/datasets. Previously, input cubes/datasets were only considered ancestors if they contributed to the shared mask. However, for lazy computations, this requires loading all the input data from disk and doing the computations before mask_fillvalues twice: once for computing the provenance and once for computing the result. This seemed too much of a computational burden for a feature that is very little used and therefore the choice was made to have less fine-grained provenance.

Needs Iris v3.9 (https://github.com/SciTools/iris/pull/5775)

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by bouweandela** [comment]
The tests pass, but the tests are not very comprehensive. Comparing data coming out `diagnostic_1/ta` of `examples/recipe_preprocessor_test.yml` with the comparison tool suggests results are different.

**Comment by bouweandela** [comment]
> Comparing data coming out diagnostic_1/ta of examples/recipe_preprocessor_test.yml with the comparison tool suggests results are different.

This has now been solved: `da.ones_like(a)` with `a` a lazy masked dask array drops the mask, while `np.ones_like(a)` with `a` a masked numpy array keeps the mask.

File: esmvalcore/preprocessor/_mask.py
Comment (schlunma) [inline esmvalcore/preprocessor/_mask.py:666]: The old code only updated ancestors for the products that actually contributed to the mask. Is this change here on purpose? I think it's fine, but we should be aware of that.

File: setup.py
Comment (schlunma) [inline setup.py:60]: You might want to update that comment here.

**Review by schlunma — COMMENTED** [review]
Thanks Bouwe, looks very good to me already! Just have some minor comments.

File: esmvalcore/preprocessor/_mask.py
Comment (bouweandela) [reply esmvalcore/preprocessor/_mask.py:666]: Thanks! I had meant to add a note to the description of the pull request, but forgot to do it. I have added it now. Does that look better?

File: setup.py
Comment (bouweandela) [reply setup.py]: Done in 9d5d3f3

File: esmvalcore/preprocessor/_mask.py
Comment (schlunma) [reply esmvalcore/preprocessor/_mask.py:666]: Great!

**Comment by schlunma** [comment]
Thanks Bouwe! Will approve once the merge conflict is solved and all tests are green 🚀

**Comment by bouweandela** [comment]
Thanks for reviewing @schlunma! Should be good to go now.

**Review by schlunma — APPROVED** [review]
🚀

**Review by valeriupredoi — APPROVED** [review]
great work folks - @bouweandela you want this in the 2.11 release or you think it can wait for later on after the release? If it can wait, pls add M2.12 to it :beer:

**Comment by bouweandela** [comment]
Thanks V! This is for the v2.12 release.

**Comment by valeriupredoi** [comment]
good man :beer:

**Pull request merged**
